### ISSUE TITLE
Director::forceSSL does not force SSL when using 

### ISSUE THREAD

**Issue by wilr** [silverstripe/silverstripe-framework#7492]
Might just be something to note in the `.env` documentation but this is something I caught in production today:

Defined a base URL in `.env`

`SS_BASE_URL="https://www.site.co.nz/"`

In `app/_config.php`: `Director::forceSSL();`.

Result:

Accessing http://www.site.co.nz incorrectly works, no redirection takes place as the check in Director::is_https() fails with the provided SS_BASE_URL.

Fix is to use the non-ssl version in the `.env` file but I think that's a simple issue for users to trip up on and could result in users being able to access pages over HTTP.

PR:

 - [ ] https://github.com/silverstripe/silverstripe-framework/pull/7520
 - [ ] https://github.com/silverstripe/silverstripe-cms/pull/2007

**Comment by dhensby** [comment]
My view is we stop recommending using forceSSL - instead use a middleware to force SSL or serverside config.

However, the problem is `Director::is_https()` check needs changing/ignoring in the `forceSSL` method as it's looking up whether the base URL is https rather than the current request.

**Comment by dhensby** [comment]
OK - so I've looked into this - calling `forceSSL()` from the `_config.php` is too early in the bootstrap process because the current `HTTPRequest` object hasn't been built yet (or isn't available for some reason).

This means that you can't make changes based on request state at this point in SS4.

So the solution is either create an SSL middleware or add this to a controller init method.

See https://github.com/silverstripe/silverstripe-framework/pull/7215#issuecomment-318360678 for some more info

**Comment by dhensby** [comment]
both forceSSL and forceWWW should be moved to middlewares or just fully deprecated in favour of server side configuration IMO

**Comment by dhensby** [comment]
I'm going to close this as "not a bug" if we have docs that recommend putting the `forceWWW` or `forceSSL` in the `_config.php` files then we need to amend those.

**Comment by wilr** [comment]
Well my 2c should be we either remove the function because it doesn't do what it says on the tin, or b we keep the function, detect that HTTPRequest hasn't been built and throw an exception as currently it's a change from 3.6. Forum is littered with use of `force_ssl`. I'm happy with either approach but we may not be able to remove for 4.0

**Comment by dhensby** [comment]
@silverstripe/core-team thoughts please?

**Comment by tractorcow** [comment]
I am in favour of throwing exception to prevent silent failures.

**Comment by sminnee** [comment]
Not keen to break APIs just now.

Would recommend that `Director::forceSSL()` is refactored to set the config value of a relevant middleware. Either it can auto-add the middleware upon calling as well (if missing), or we can include the relevant middleware by default. It can throw a deprecation warning if we like.

Then the middleware can do the hard work, as Dan suggests.

I'd probably design the middleware to be a little more generic than just forceSSL: cover force ssl, add www, canonicalise domain name.

**Comment by tractorcow** [comment]
Maybe not setting a config value (setting config isn't recommended anymore) but yes some flag that triggers the later middleware would be great :)

That way forceSSL() in _config.php would work normally.

**Comment by sminnee** [comment]
Setting config isn't recommended, but neither is `Director::forceSSL()`, so I think that's okay ;-)

**Comment by tractorcow** [comment]
Yes we can deprecate it with a new middleware at the same time. :)

**Comment by dhensby** [comment]
OK - are we happy to add new middleware now we are in RC or will this change have to wait?

**Comment by tractorcow** [comment]
I think 4.1 is fine.

**Comment by wilr** [comment]
IMHO it should be `4.0` or at least marked up in bold in the changelog as something to replace as if `Director::forceSSL` no longer works I can name at least a dozen govt agencies where we're used that rather than in `.htaccess` since they use different domains.

**Comment by sminnee** [comment]
Marked as a high bug because SSL matters.

**Comment by sminnee** [comment]
@wilr 

>  I can name at least a dozen govt agencies where we're used that rather than in .htaccess since they use different domains.

Do you mean that they have multiple certificates / domain names that their site operates under, or is there a single canonical HTTPS domain that everyone should be redirected to? Just wondering to help flesh out how we structure the undeprecated API.

**Comment by tractorcow** [comment]
Ok, let's fix in rc2?

**Comment by stojg** [comment]
From an operational and deployment perspective I would like to mention that any redirection logic should not be triggered in CLI environments. This has in the past been surfaced bugs where for example it's been triggered on dev/builds or other tasks with the result of a sudden exit without any output.

**Comment by tractorcow** [comment]
I believe in CLI the expected behaviour is that it would immediately process the redirected URL. It's not consistently implemented though. We may want to adjust our expectations here. :)

**Comment by stojg** [comment]
Yeah, absolutely, just want to make sure we all realise there are 2 supported environments for SilverStripe :)

**Comment by stojg** [comment]
SSL redirection is a specialised case of redirection, other cases are redirecting to a specific domain or subdomain (domain.com -> www.domain.com, ie domain canonicalization).

If you are running several websites on a server you would then need to set up multiple virtual hosts with a list of listening domains.

If then SilverStripe does the redirection, you would need to synchronise the domains and redirections between the SilverStripe application and the web server / forward proxies.

The above might be a bit more complicated case, but I think it highlights that it's easy to cause unexpected bugs in production mode that cannot easily be reproduced in dev / UAT environments.

For simple setups, where the web server just points to a document root for all incoming connections, it's easy enough for SilverStripe to take care of all redirection/domain checking. 

I don't have a good answer on where the responsibilities of a web server end and SilverStripes begin.

**Comment by tractorcow** [comment]
I would propose expected behaviour for forceSSL() in CLI is that the request was processed where is_https() would return true. It could be as simple as modifying the protocol of the current request object, or otherwise it could re-start a request with this flag set.

This would be necessary for code paths which generate links based on the "current url", which should then have the necessary `https://` prefix.

**Comment by tractorcow** [comment]
Fix at https://github.com/silverstripe/silverstripe-framework/pull/7520

For #7509 we can just never run that middleware (it'd be disabled).

**Comment by sminnee** [comment]
Added to 7509 "Note that the implication of this is that no HTTPMiddlewares would ever get run in CLI mode, which simplifies things such as #7492."

**Comment by jmfederico** [comment]
I just noticed that when running "sake dev/build" from the command line, the base_path will always be http, even though I have https in `SS_BASE_URL`.

This is a problem when generating static files (like error-404.html).

I read this issue and #7509 but I am not sure if this has been solved, or if there is a solution I can apply to my project (apart from running dev/build from a browser instead.)

I am guessing this is the same issue, but I am not sure.

Is this something you knew already?
